source/Halibut.Tests/Support/BackwardsCompatibility/ProxyHalibutTestBinaryRunner.cs

```diff
@@ -82,38 +82,54 @@ public async Task<RoundTripRunningOldHalibutBinary> Run()
             {
                 settings.Add("realServiceListenAddress", realServiceListenAddress.ToString());
             }
+            
+            var tmpDirectory = new TmpDirectory();
 
-            var cts = new CancellationTokenSource();
+            var (task, serviceListenPort, proxyClientListenPort, runningTentacleCancellationTokenSource) = await StartHalibutTestBinary(version, settings, tmpDirectory);
 
-            try
-            {
-                var tmp = new TmpDirectory();
-
-                var (task, serviceListenPort, proxyClientListenPort) = await StartHalibutTestBinary(version, settings, tmp, cts.Token);
-
-                return new RoundTripRunningOldHalibutBinary(cts, task, tmp, serviceListenPort, proxyClientListenPort, compatBinaryStayAlive);
-            }
-            catch (Exception)
-            {
-                cts.Cancel();
-                cts.Dispose();
-                throw;
-            }
+            return new RoundTripRunningOldHalibutBinary(runningTentacleCancellationTokenSource, task, tmpDirectory, serviceListenPort, proxyClientListenPort, compatBinaryStayAlive);
         }
 
-        async Task<(Task, int?, int?)> StartHalibutTestBinary(string version, Dictionary<string, string> settings, TmpDirectory tmp, CancellationToken cancellationToken)
+        async Task<(Task RunningTentacleTask, int? ServiceListenPort, int? ProxyClientListenPort, CancellationTokenSource RunningTentacleCancellationTokenSource)> StartHalibutTestBinary(
```

Comment:
I've refactored the way cancellation works so it's a bit more aggressive to try and clean things up. I doubt this is the underlying cause of hangs but just incase